src/header-parser.ts

```diff
@@ -0,0 +1,336 @@
+import type Spec from './Spec';
+
+export function parseStructuredHeaderH1(
+  spec: Spec,
+  header: Element
+): { name: string | null; formattedHeader: string | null; formattedParams: string | null } {
+  // parsing is intentionally permissive; the linter can do stricter checks
+  // TODO have the linter do checks
+  let headerText = header.innerHTML;
+  const prefix = headerText.match(/^\s*(Static|Runtime) Semantics:\s*/);
+  if (prefix != null) {
+    headerText = headerText.substring(prefix[0].length);
+  }
+
+  let formattedHeader = null;
+  let formattedParams = null;
+
+  const parsed = headerText.match(/^\s*([^(\s]+)\s*\((.*)\)\s*$/s);
+  if (parsed == null) {
+    spec.warn({
+      type: 'contents',
+      ruleId: 'header-format',
+      message: `failed to parse header`,
+      node: header,
+      nodeRelativeColumn: 1,
+      nodeRelativeLine: 1,
+    });
+    return { name: null, formattedHeader, formattedParams };
+  }
+
+  type Param = { name: string; type: string | null };
+  const name = parsed[1];
+  let paramText = parsed[2].trim();
+  const params: Array<Param> = [];
+  const optionalParams: Array<Param> = [];
+
+  if (/\(\s*\n/.test(headerText)) {
+    // if it's multiline, parse it for types
+    const paramChunks = paramText
+      .split('\n')
+      .map(c => c.trim())
+      .filter(c => c.length > 1);
+    let index = 0;
+    for (const chunk of paramChunks) {
+      ++index;
+      // TODO linter enforces all optional params come after non-optional params
+      const parsedChunk = chunk.match(/^(optional\s*)?([A-Za-z0-9_]+)\s*:\s*(\S.*\S)/);
+      if (parsedChunk == null) {
+        spec.warn({
+          type: 'contents',
+          ruleId: 'header-format',
+          message: `failed to parse header (parameter ${index})`,
+          node: header,
+          nodeRelativeColumn: 1,
+          nodeRelativeLine: 1,
```

Comment:
We could pretty easily point to the correct line here by zipping the split/trim result with the naturals. I think it might be pretty useful since this is possibly quite a bit of text.

Reply:
It's more complicated than that, but done.